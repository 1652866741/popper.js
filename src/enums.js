// @flow
export const top: 'top' = 'top';
export const bottom: 'bottom' = 'bottom';
export const right: 'right' = 'right';
export const left: 'left' = 'left';
export type BasePlacement =
  | typeof top
  | typeof bottom
  | typeof right
  | typeof left;
export const basePlacements: Array<BasePlacement> = [top, bottom, right, left];

export type Placement =
  | 'top'
  | 'top-start'
  | 'top-end'
  | 'bottom'
  | 'bottom-start'
  | 'bottom-end'
  | 'right'
  | 'right-start'
  | 'right-end'
  | 'left'
  | 'left-start'
  | 'left-end';

export const placements: Array<Placement> = basePlacements.reduce(
  (acc: Array<Placement>, placement: BasePlacement): Array<Placement> =>
    acc.concat([
      placement,
      (`${placement}-start`: any),
      (`${placement}-end`: any),
    ]),
  []
);
