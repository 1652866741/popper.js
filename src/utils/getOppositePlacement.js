/**
 * Get the opposite placement of the given one/
 * @function
 * @ignore
 * @argument {String} placement
 * @returns {String} flipped placement
 */
export default function getOppositePlacement(placement) {
    const hash = {left: 'right', right: 'left', bottom: 'top', top: 'bottom' };
    return placement.replace(/left|right|bottom|top/g, (matched) => hash[matched]);
}
