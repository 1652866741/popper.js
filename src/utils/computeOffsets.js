// @flow
import getBasePlacement from './getBasePlacement';
import getMainAxisFromPlacement from './getMainAxisFromPlacement';
import getAltAxis from './getAltAxis';
import getAltLen from './getAltLen';
import type { Rect, PositioningStrategy, Offsets } from '../types';
import { top, right, bottom, left, type Placement } from '../enums';

export default ({
  reference,
  popper,
  strategy,
  placement,
}: {
  reference: Rect,
  popper: Rect,
  strategy: PositioningStrategy,
  placement: Placement,
}): Offsets => {
  const basePlacement = getBasePlacement(placement);

  switch (basePlacement) {
    case top:
      return {
        x: reference.x + reference.width / 2,
        y: reference.y - reference.height,
      };
    case bottom:
      return {
        x: reference.x + popper.width / 2,
        y: reference.y + reference.height,
      };
    case right:
      return {
        x: reference.x + reference.width,
        y: reference.y + popper.height / 2,
      };
    case left:
      return {
        x: reference.x - reference.width,
        y: reference.y + popper.height / 2,
      };
    default:
      // $FlowFixMe: This will actually never match github.com/facebook/flow/issues/2395
      return undefined;
  }
};
