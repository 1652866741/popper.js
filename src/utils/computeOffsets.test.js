// @flow
import createRectMock from 'createRectMock';
import computeOffsets from './computeOffsets';
import { basePlacements, placements } from '../enums';
import type { Placement } from '../enums';
import getMainAxisFromPlacement from './getMainAxisFromPlacement';
import getAltAxis from './getAltAxis';
import getOppositePlacement from './getOppositePlacement';

const reference = createRectMock({
  width: 200,
  height: 200,
  x: 100,
  y: 100,
});
const popper = createRectMock({ width: 100, height: 100, x: 0, y: 0 });

describe('[strategy: absolute]', () =>
  basePlacements.forEach((placement: Placement) =>
    describe('[placement: ${placement}]', () => {
      it('computes offsets', () => {
        expect(
          computeOffsets({
            reference,
            popper,
            strategy: 'absolute',
            placement,
          })
        ).toMatchSnapshot('x and y');

        expect(
          computeOffsets({
            reference,
            popper,
            strategy: 'absolute',
            placement,
          })[getMainAxisFromPlacement(placement)]
        ).toEqual(
          computeOffsets({
            reference,
            popper,
            strategy: 'absolute',
            placement: getOppositePlacement(placement),
          })[getMainAxisFromPlacement(placement)]
        );
      });
    })
  ));
