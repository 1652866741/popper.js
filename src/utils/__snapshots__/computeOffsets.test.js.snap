// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[strategy: absolute] [placement: \${placement}] computes offsets: x and y 1`] = `
Object {
  "x": 150,
  "y": 0,
}
`;

exports[`[strategy: absolute] [placement: \${placement}] computes offsets: x and y 2`] = `
Object {
  "x": 150,
  "y": 200,
}
`;

exports[`[strategy: absolute] [placement: \${placement}] computes offsets: x and y 3`] = `
Object {
  "x": 200,
  "y": 150,
}
`;

exports[`[strategy: absolute] [placement: \${placement}] computes offsets: x and y 4`] = `
Object {
  "x": 0,
  "y": 150,
}
`;
